File: monty hall/monty-hall.py
import random

def monty_hall(rounds, switch):
    # Initialize counters for wins and losses
    wins = 0
    losses = 0

    for round_num in range(1, rounds + 1):
        # Step 1: Randomly place the car behind one of the doors
        doors = [0, 0, 0]  # 0 means no car, 1 means car
        car_position = random.randint(0, 2)
        doors[car_position] = 1
        
        # Step 2: Automatically choose a random door (simulate user's choice)
        user_choice = random.randint(0, 2)  # Random choice between door 0, 1, or 2
        print(f"Round {round_num} - Your automatic choice is door {user_choice + 1}")
        
        # Step 3: Reveal a wrong door
        possible_reveal = [i for i in range(3) if i != user_choice and doors[i] != 1]
        revealed_door = random.choice(possible_reveal)
        print(f"I reveal door {revealed_door + 1} and there is no car behind it.")
        
        # Step 4: If the user is switching, change their choice
        if switch:
            user_choice = 3 - user_choice - revealed_door  # The remaining door is the switch option
        
        # Step 5: Reveal if the user won
        if doors[user_choice] == 1:
            print("Congratulations! You won the car!")
            wins += 1
        else:
            print("Sorry, no car behind that door. Better luck next time!")
            losses += 1
        
        print(f"Current score: {wins} Wins, {losses} Losses\n")
    
    # Final results after all rounds
    print(f"Game Over! Final score: {wins} Wins, {losses} Losses")

if __name__ == "__main__":
    print("How many rounds would you like to play?")
    rounds = int(input())  # Input for number of rounds
    
    print("Do you want to switch your answer every time? (yes or no)")
    switch_choice = input().lower()
    
    # Convert input to boolean
    switch = switch_choice == "yes"
    
    monty_hall(rounds, switch)

File: monty hall/monty_hall.js
const readline = require("readline");

// Create an interface for user input
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
});

function montyHall(rounds, switchChoice) {
    let wins = 0;
    let losses = 0;
    let completedRounds = 0;

    const batchSize = 1000000; // Process in batches of 1 million
    const totalBatches = Math.ceil(rounds / batchSize);

    console.log(`Starting simulation of ${rounds.toLocaleString()} rounds...`);

    for (let batch = 0; batch < totalBatches; batch++) {
        const currentBatchSize = Math.min(batchSize, rounds - completedRounds);

        for (let i = 0; i < currentBatchSize; i++) {
            // Step 1: Randomly place the car behind one of the doors
            const carPosition = Math.floor(Math.random() * 3);

            // Step 2: Randomly select a door
            let userChoice = Math.floor(Math.random() * 3);

            // Step 3: Reveal a wrong door
            const possibleReveals = [0, 1, 2].filter(
                (door) => door !== userChoice && door !== carPosition
            );
            const revealedDoor = possibleReveals[Math.floor(Math.random() * possibleReveals.length)];

            // Step 4: If switching, change the user's choice
            if (switchChoice) {
                userChoice = [0, 1, 2].find(
                    (door) => door !== userChoice && door !== revealedDoor
                );
            }

            // Step 5: Check if the user wins
            if (userChoice === carPosition) {
                wins++;
            } else {
                losses++;
            }
        }

        completedRounds += currentBatchSize;

        // Log progress after each batch
        console.log(`${completedRounds.toLocaleString()} rounds completed...`);
    }

    // Output final results
    console.log(`Simulation completed.`);
    console.log(`Rounds Played: ${rounds.toLocaleString()}`);
    console.log(`Wins: ${wins.toLocaleString()}`);
    console.log(`Losses: ${losses.toLocaleString()}`);
    console.log(`Win Rate: ${(wins / rounds * 100).toFixed(6)}%`);
}

// Prompt user for inputs
rl.question("How many rounds would you like to play? ", (roundsInput) => {
    const rounds = parseInt(roundsInput);

    rl.question("Do you want to switch your choice every time? (yes/no) ", (switchInput) => {
        const switchChoice = switchInput.toLowerCase() === "yes";

        console.time("MontyHall Simulation");
        montyHall(rounds, switchChoice);
        console.timeEnd("MontyHall Simulation");

        rl.close();
    });
});
